### PR TITLE
Fixing keyboard interrupts on posix OSes

### PR DESCRIPTION
**Pull request by bridadan**
This should fix issue #150. Tested working on Windows 10 and Ubuntu Linux 16.04.

The underlying issue was that a blocking call to `join` for the worker threads was stopping `KeyboardInterrupt` exceptions from being raised to the exception handler here: https://github.com/ARMmbed/greentea/blob/master/mbed_greentea/mbed_greentea_cli.py#L396

This fix changes the blocking call to `join` to one that timesout after 0.1 seconds. The join is called in a loop until the thread's `isAlive()` function returns False.

@geky @theotherjimmy @jupe I'm sure you'll be very happy to see this fix finally happen 😄 

@mazimkhan Could you please take a look?

### CODE REVIEW

File: mbed_greentea/mbed_greentea_cli.py [hunk 1]
```diff
@@ -906,7 +906,12 @@ def get_parallel_value(value):
         # merge partial test reports from different threads to final test report
         for t in execute_threads:
             try:
-                t.join() #blocking
+                # We can't block forever here since that prevents KeyboardInterrupts
+                # from being propagated correctly. Therefore, we just join with a
+                # timeout of 0.1 seconds until the thread isn't alive anymore.
+                while t.isAlive():
+                    t.join(0.1)
```
Comment (adbridge) [inline mbed_greentea/mbed_greentea_cli.py]: Any particular reason for choosing 0.1 secs ?

**Comment by mazimkhan** [comment]
Looks good to me. Though isn't 100ms too aggressive for an unlikely case.

**Comment by bridadan** [comment]
The `0.1` seconds was arbitrary I admit. It just needs to balance CPU utilization and responsiveness to the users Ctrl+C keyboard stroke. This should be a very manageable load on the CPU (on Windows it was using <1% during testing). It should also still feel very responsive (vs a 1 second poll period for instance).

> Though isn't 100ms too aggressive for an unlikely case.

@mazimkhan Not sure what the aggressive case is, can you explain?

**Comment by adbridge** [comment]
@bridadan Might be worth adding something along those lines to the code

**Comment by bridadan** [comment]
@adbridge comments amended!

**Comment by adbridge** [comment]
@mazimkhan are you happy with this?

**Comment by mazimkhan** [comment]
Yup looks goods.

**Comment by adbridge** [comment]
@bridadan did you test this worked on other OS' ? I tried in in the mbed2 release script running on gitbash (windows) and it failed unless I also had 
        thr.setDaemon(True)
for each thread...

**Comment by bridadan** [comment]
@adbridge I tested on Windows 10 and Ubuntu Linux 16.04. What OS are you testing on? It may be that we have to add that as well. Please let me know how to reproduce!

**Comment by adbridge** [comment]
@bridadan Shell = gitbash, running on windows 10 . I had multiple concurrent threads in the mbed2 release script so I don't know if that makes a difference either.

**Comment by bridadan** [comment]
@adbridge Sorry I'm still not 100% sure what you mean. Are you saying that when you were implementing multithreading in the release script that you had to use `thread.setDaemon(True)` for it to work? If so, it looks like we're already setting these threads to be daemon threads: https://github.com/bridadan/greentea/blob/bf221e28ed1d9a00f0179140e854309d7ad9a018/mbed_greentea/mbed_greentea_cli.py#L903

Or did you try this PR locally and it still didn't work?

**Comment by adbridge** [comment]
@bridadan the former :) . I only looked at the actual patch not what was there previously , so if the threads were already set up as daemons then you're probably fine! :)

**Pull request merged**